File: Examples/MODAV SwiftUI/MODAV SwiftUI/Orchestrator/UserOrchestrator.swift
//
//  UserOrchestrator.swift
//  MODAV SwiftUI
//
//  Created by Emin Emini on 17.12.2023..
//

import Foundation

// MARK: - Orchestrator
class UserOrchestrator {
    weak var delegate: UserDelegate?
    private let adapter = UserAPIManager()
    
    func getUsers(count: Int) {
        UserAPIManager.shared.getUsers(count: count) { result in
            switch result {
            case .success(let users):
                // Handle success with users array
                self.delegate?.didFetchedSuccessfully(users: users)
            case .failure(let error):
                // Handle error
                self.delegate?.didFailWithError(error: error)
            }
        }

    }
}

File: Examples/MODAV-Kotlin/app/src/main/java/com/eminemini/modavkotlin/UserOrchestrator.kt
package com.eminemini.modavkotlin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class UserOrchestrator {
    var delegate: UserDelegate? = null
    private val adapter = UserAPIManager()

    fun getUsers(count: Int) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val users = adapter.getUsers(count)
                CoroutineScope(Dispatchers.Main).launch {
                    delegate?.didFetchedSuccessfully(users)
                }
            } catch (e: Exception) {
                CoroutineScope(Dispatchers.Main).launch {
                    delegate?.didFailWithError(e)
                }
            }
        }
    }
}